src/algebra/pointwise.lean

```diff
@@ -157,151 +140,180 @@ begin
+| ⟨x, hx⟩ ⟨y, hy⟩ := ⟨x * y, ⟨x, y, hx, hy, rfl⟩⟩
 
 @[simp, to_additive]
-lemma univ_pointwise_mul_univ [monoid α] : (univ : set α) * univ = univ :=
+lemma univ_mul_univ [monoid α] : (univ : set α) * univ = univ :=
 begin
-  have : ∀x, ∃a b : α, x = a * b := λx, ⟨x, ⟨1, (mul_one x).symm⟩⟩,
-  show {a | ∃ x ∈ univ, ∃ y ∈ univ, a = x * y} = univ,
+  have : ∀x, ∃a b : α, a * b = x := λx, ⟨x, ⟨1, mul_one x⟩⟩,
+  show {a | ∃ x y, x ∈ univ ∧ y ∈ univ ∧ x * y = a} = univ,
   simpa [eq_univ_iff_forall]
 end
 
-def pointwise_mul_fintype [has_mul α] [decidable_eq α] (s t : set α) [hs : fintype s] [ht : fintype t] :
-  fintype (s * t : set α) := by { rw pointwise_mul_eq_image, apply set.fintype_image }
+@[to_additive]
+def fintype_mul [has_mul α] [decidable_eq α] (s t : set α)
+  [hs : fintype s] [ht : fintype t] : fintype (s * t : set α) :=
+by { rw ← mul_image_prod, apply set.fintype_image }
 
-def pointwise_add_fintype [has_add α] [decidable_eq α] (s t : set α) [hs : fintype s] [ht : fintype t] :
-  fintype (s + t : set α) := by { rw pointwise_add_eq_image, apply set.fintype_image }
+@[to_additive set.has_neg'] -- todo: remove prime once name becomes available
+instance [has_inv α] : has_inv (set α) :=
+⟨preimage has_inv.inv⟩
 
-attribute [to_additive] set.pointwise_mul_fintype
+@[simp, to_additive]
+lemma mem_inv [has_inv α] : a ∈ s⁻¹ ↔ a⁻¹ ∈ s := iff.rfl
 
-/-- Pointwise scalar multiplication by a set of scalars. -/
-def pointwise_smul [has_scalar α β] : has_scalar (set α) (set β) :=
-  ⟨λ s t, { x | ∃ a ∈ s, ∃ y ∈ t, x  = a • y }⟩
+@[to_additive]
+lemma inv_mem_inv [group α] : a⁻¹ ∈ s⁻¹ ↔ a ∈ s :=
+by simp only [mem_inv, inv_inv]
 
-/-- Scaling a set: multiplying every element by a scalar. -/
-def smul_set [has_scalar α β] : has_scalar α (set β) :=
-  ⟨λ a s, { x | ∃ y ∈ s, x = a • y }⟩
+@[simp, to_additive]
+lemma inv_preimage [has_inv α] : has_inv.inv ⁻¹' s = s⁻¹ :=
+rfl
 
-local attribute [instance] pointwise_smul smul_set
+@[simp, to_additive]
+lemma inv_image [group α] : has_inv.inv '' s = s⁻¹ :=
+by { simp only [← inv_preimage], rw [image_eq_preimage_of_inverse]; intro; simp only [inv_inv] }
 
-lemma mem_smul_set [has_scalar α β] (a : α) (s : set β) (x : β) :
-  x ∈ a • s ↔ ∃ y ∈ s, x = a • y := iff.rfl
+@[simp, to_additive]
+lemma inter_inv [has_inv α] : (s ∩ t)⁻¹ = s⁻¹ ∩ t⁻¹ :=
+preimage_inter
 
-lemma smul_set_eq_image [has_scalar α β] (a : α) (s : set β) :
-  a • s = (λ x, a • x) '' s :=
-set.ext $ λ x, iff.intro
-  (λ ⟨_, hy₁, hy₂⟩, ⟨_, hy₁, hy₂.symm⟩)
-  (λ ⟨_, hy₁, hy₂⟩, ⟨_, hy₁, hy₂.symm⟩)
+@[simp, to_additive]
+lemma union_inv [has_inv α] : (s ∪ t)⁻¹ = s⁻¹ ∪ t⁻¹ :=
+preimage_union
 
-lemma smul_set_eq_pointwise_smul_singleton [has_scalar α β]
-  (a : α) (s : set β) : a • s = ({a} : set α) • s :=
-set.ext $ λ x, iff.intro
-  (λ ⟨_, h⟩, ⟨a, mem_singleton _, _, h⟩)
-  (λ ⟨_, h, y, hy, hx⟩, ⟨_, hy, by {
-    rw mem_singleton_iff at h; rwa h at hx }⟩)
+@[simp, to_additive]
+lemma compl_inv [has_inv α] : (s.compl)⁻¹ = (s⁻¹).compl :=
```

Comment:
Should this now use the new compl notation?

Reply:
Yes, it should. `set.compl` should not be used.